### PR TITLE
Try optimise relation writing - Part 2

### PR DESCRIPTION
**Pull request by fppt**
More aggressive caching. This is a bit experimental

### CODE REVIEW

File: grakn-kb/src/main/java/ai/grakn/kb/internal/ValidateGlobalRules.java [hunk 1]
```diff
@@ -314,12 +314,12 @@ private ValidateGlobalRules() {
...
-     * @param relationshipReified the {@link Relationship} to check
+     * @param relationReified the {@link Relationship} to check
      * @return An error message if the {@link Relationship} is not unique.
      */
-    private static Optional<String> validateNonKeyControlledRelation(RelationshipReified relationshipReified){
-        String hash = RelationshipReified.generateNewHash(relationshipReified.type(), relationshipReified.allRolePlayers());
-        return setRelationUnique(relationshipReified, hash);
+    private static Optional<String> validateNonKeyControlledRelation(RelationshipReified relationReified){
+        String hash = RelationshipReified.generateNewHash(relationReified.type(), relationReified.allRolePlayers());
+        return setRelationUnique(relationReified, hash);
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/ValidateGlobalRules.java]: ok now you're just trying to upset me

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipReified.java [hunk 2]
```diff
@@ -58,6 +65,19 @@
...
+     * Set of {@link Casting}s which are loaded into memory. This should only be used when adding new roleplayers.
+     * This is because validation requires iterating over all castings anyway.
+     */
+    private Cache<Set<Casting>> allCastings = new Cache<>(Cacheable.set(), () -> {
+        return vertex().tx().getTinkerTraversal().V().
+                has(Schema.VertexProperty.ID.name(), getId().getValue()).
+                outE(Schema.EdgeLabel.ROLE_PLAYER.getLabel()).
+                has(Schema.EdgeProperty.RELATIONSHIP_TYPE_LABEL_ID.name(), type().getLabelId().getValue()).
+                toStream().map(edge -> vertex().tx().factory().buildCasting(edge)).
+                collect(Collectors.toSet());
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipReified.java]: Any reason we can't just do `vertex().getEdgesOfType(Direction.OUT, Schema.EdgeLabel.ROLE_PLAYER)`?

File: grakn-kb/src/main/java/ai/grakn/kb/internal/ValidateGlobalRules.java (hunk 1 above)
Comment (fppt) [reply grakn-kb/src/main/java/ai/grakn/kb/internal/ValidateGlobalRules.java]: oooops merge error

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipReified.java (hunk 2 above)
Comment (fppt) [reply grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipReified.java]: If I did that then the index is ignored. I will put a comment there.

**Pull request merged**